core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,406 @@
+        RebalanceContext context = new RebalanceContext(writerPhysicalWrittenBytes, partitionsRowCount);
+
+        IndexedPriorityQueue<WriterId> maxWriters = new IndexedPriorityQueue<>();
+        IndexedPriorityQueue<WriterId> minWriters = new IndexedPriorityQueue<>();
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            WriterId writer = new WriterId(writerId);
+            maxWriters.addOrUpdate(writer, context.getWriterPriority(writer));
+            minWriters.addOrUpdate(writer, Long.MAX_VALUE - context.getWriterPriority(writer));
+        }
+
+        // Find skewed partitions and scale them across multiple writers
+        while (true) {
+            // Find the writer with maximum physical written bytes since last rebalance
+            WriterId maxWriter = maxWriters.poll();
+
+            if (maxWriter == null) {
+                break;
+            }
+
+            // Find the skewness against writer with max physical written bytes since last rebalance
+            List<WriterId> minSkewedWriters = findSkewedMinWriters(context, maxWriter, minWriters);
+            if (minSkewedWriters.isEmpty()) {
+                break;
+            }
+
+            for (WriterId minSkewedWriter : minSkewedWriters) {
+                if (context.rebalancePartition(maxWriter, minSkewedWriter)) {
+                    maxWriters.addOrUpdate(maxWriter, context.getWriterPriority(maxWriter));
+                    minWriters.addOrUpdate(maxWriter, Long.MAX_VALUE - context.getWriterPriority(maxWriter));
+                    break;
+                }
+            }
+
+            // Add all the min skewed writers back to the minWriters queue with updated priorities
+            for (WriterId minSkewedWriter : minSkewedWriters) {
+                minWriters.addOrUpdate(minSkewedWriter, Long.MAX_VALUE - context.getWriterPriority(minSkewedWriter));
+            }
+        }
+
+        resetStateForNextRebalance(context, writerPhysicalWrittenBytes, partitionsRowCount);
+    }
+
+    private int getPhysicalWrittenBytesSinceLastRebalance(List<Long> writerPhysicalWrittenBytes)
+    {
+        int physicalWrittenBytesSinceLastRebalance = 0;
+        for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+            physicalWrittenBytesSinceLastRebalance +=
+                    writerPhysicalWrittenBytes.get(writerId) - writerPhysicalWrittenBytesAtLastRebalance.get(writerId);
+        }
+
+        return physicalWrittenBytesSinceLastRebalance;
+    }
+
+    private List<WriterId> findSkewedMinWriters(RebalanceContext context, WriterId maxWriter, IndexedPriorityQueue<WriterId> minWriters)
+    {
+        ImmutableList.Builder<WriterId> minSkewedWriters = ImmutableList.builder();
+        long maxWriterWrittenBytes = context.getWriterPriority(maxWriter);
+        while (true) {
+            // Find the writer with minimum written bytes since last rebalance
+            WriterId minWriter = minWriters.poll();
```

Comment:
Why not use iterator from `IndexedPriorityQueue` instead of while + poll and `null` check?

Reply:
`poll` also removes the element from the queue and index which is important.